voip_utils/sip.py

```diff
@@ -283,3 +349,280 @@ def _parse_sip(
         body = message[offset:]
 
         return method, ruri, headers, body
+
+
+class CallPhoneDatagramProtocol(asyncio.DatagramProtocol, ABC):
+    def __init__(
+        self,
+        sdp_info: SdpInfo | None,
+        source: SipEndpoint,
+        dest: SipEndpoint,
+        rtp_port: int,
+        loop: Optional[asyncio.AbstractEventLoop] = None,
+    ) -> None:
+        self.sdp_info = sdp_info
+        self.transport = None
+        self._closed_event = asyncio.Event()
+        self._loop = loop if loop is not None else asyncio.get_running_loop()
+        self._session_id = str(time.monotonic_ns())
+        self._session_version = str(time.monotonic_ns())
+        self._call_id = str(time.monotonic_ns())
+        self._source_endpoint = source
+        self._dest_endpoint = dest
+        self._rtp_port = rtp_port
+        self._request_uri = f"sip:{dest.username}@{dest.host}:{dest.port}"
+
+    def connection_made(self, transport):
+        self.transport = transport
+
+        sdp_lines = [
+            "v=0",
+            f"o={self._source_endpoint.username} {self._session_id} {self._session_version} IN IP4 {self._source_endpoint.host}",
+            "s=Talk",
+            f"c=IN IP4 {self._source_endpoint.host}",
+            "t=0 0",
+            f"m=audio {self._rtp_port} RTP/AVP 123 96 101 103 104",
+            "a=sendrecv",
+            "a=rtpmap:96 opus/48000/2",
+            "a=fmtp:96 useinbandfec=0",
+            "a=rtpmap:123 opus/48000/2",
+            "a=fmtp:123 maxplaybackrate=16000",
+            "a=rtpmap:101 telephone-event/48000",
+            "a=rtpmap:103 telephone-event/16000",
+            "a=rtpmap:104 telephone-event/8000",
+            "a=ptime:20",
+            "",
+        ]
+        sdp_text = _CRLF.join(sdp_lines)
+        sdp_bytes = sdp_text.encode("utf-8")
+
+        invite_lines = [
+            f"INVITE {self._request_uri} SIP/2.0",
+            f"Via: SIP/2.0/UDP {self._source_endpoint.host}:{self._source_endpoint.port}",
+            f"From: <sip:{self._source_endpoint.username}@{self._source_endpoint.host}:{self._source_endpoint.port}>",
+            f"Contact: <sip:{self._source_endpoint.username}@{self._source_endpoint.host}:{self._source_endpoint.port}>",
+            f"To: <sip:{self._dest_endpoint.username}@{self._dest_endpoint.host}:{self._dest_endpoint.port}>",
+            f"Call-ID: {self._call_id}",
+            "CSeq: 50 INVITE",
+            "User-Agent: test-agent 1.0",
+            "Allow: INVITE, ACK, OPTIONS, CANCEL, BYE, SUBSCRIBE, NOTIFY, INFO, REFER, UPDATE",
+            "Accept: application/sdp, application/dtmf-relay",
+            "Content-Type: application/sdp",
+            f"Content-Length: {len(sdp_bytes)}",
+            "",
+        ]
+        invite_text = _CRLF.join(invite_lines) + _CRLF
+        invite_bytes = invite_text.encode("utf-8")
+
+        print((invite_bytes + sdp_bytes).decode())
+
```

Comment:
Debug print still left in the code. If you want to keep this, consider logger. Also you can consider logging sdp_text and invite_text directly to avoid a decoding step.